components/browser/search/src/main/java/mozilla/components/browser/search/suggestions/SearchSuggestionClient.kt

```diff
@@ -0,0 +1,38 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.search.suggestions
+
+import mozilla.components.browser.search.SearchEngine
+
+/**
+ * Async function responsible for taking a URL and returning the results
+ */
+typealias SearchSuggestionFetcher = suspend (url: String) -> String?
+
+/**
+ *  Provides an interface to get search suggestions from a given SearchEngine.
+ */
+class SearchSuggestionClient private constructor(
+    private val searchEngine: SearchEngine,
+    private val fetcher: SearchSuggestionFetcher
+) {
+    companion object {
+        fun create(searchEngine: SearchEngine, fetcher: SearchSuggestionFetcher): SearchSuggestionClient? {
+            if (!searchEngine.canProvideSearchSuggestions) { return null }
+
+            return SearchSuggestionClient(searchEngine, fetcher)
+        }
+    }
```

Comment:
Do we really need this factory method?

You can just throw from init:

```Kotlin
class SearchSuggestionClient private constructor(
    private val searchEngine: SearchEngine,
    private val fetcher: SearchSuggestionFetcher
) {
    init {
        if (!searchEngine.canProvideSearchSuggestions) {
           throw IllegalStateException(...)
       }
    }
}
```

It also avoid that you have to deal with `null` values coming out of `create()`. Especially if we make `canProvideSearchSuggestions` public then we do not need to try to create a client for this search engine in the first place.

Reply:
ah! I totally forgot about inits with kotlin 🤦‍♂️